File: qc-application/src/main/java/com/eager/questioncloud/application/api/hub/review/service/HubReviewService.java
package com.eager.questioncloud.application.api.hub.review.service;

import com.eager.questioncloud.application.api.hub.review.implement.HubReviewRegister;
import com.eager.questioncloud.application.api.hub.review.implement.HunReviewUpdater;
import com.eager.questioncloud.core.common.PagingInformation;
import com.eager.questioncloud.core.domain.review.dto.MyQuestionReview;
import com.eager.questioncloud.core.domain.review.dto.QuestionReviewDetail;
import com.eager.questioncloud.core.domain.review.event.DeletedReviewEvent;
import com.eager.questioncloud.core.domain.review.event.ModifiedReviewEvent;
import com.eager.questioncloud.core.domain.review.event.RegisteredReviewEvent;
import com.eager.questioncloud.core.domain.review.infrastructure.QuestionReviewRepository;
import com.eager.questioncloud.core.domain.review.model.QuestionReview;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class HubReviewService {
    private final QuestionReviewRepository questionReviewRepository;
    private final HubReviewRegister hubReviewRegister;
    private final HunReviewUpdater questionHubReviewUpdater;
    private final ApplicationEventPublisher applicationEventPublisher;

    public int getTotal(Long questionId) {
        return questionReviewRepository.getTotal(questionId);
    }

    public List<QuestionReviewDetail> getQuestionReviews(Long questionId, Long userId, PagingInformation pagingInformation) {
        return questionReviewRepository.getQuestionReviews(questionId, userId, pagingInformation);
    }

    public MyQuestionReview getMyQuestionReview(Long questionId, Long userId) {
        QuestionReview questionReview = questionReviewRepository.getMyQuestionReview(questionId, userId);
        return MyQuestionReview.from(questionReview);
    }

    public void register(QuestionReview questionReview) {
        hubReviewRegister.register(questionReview);
        applicationEventPublisher.publishEvent(RegisteredReviewEvent.create(questionReview.getQuestionId(), questionReview.getRate()));
    }

    public void modify(Long reviewId, Long userId, String comment, int rate) {
        QuestionReview questionReview = questionReviewRepository.findByIdAndUserId(reviewId, userId);
        int varianceRate = questionHubReviewUpdater.modifyQuestionReview(questionReview, comment, rate);
        applicationEventPublisher.publishEvent(ModifiedReviewEvent.create(questionReview.getQuestionId(), varianceRate));
    }

    public void delete(Long reviewId, Long userId) {
        QuestionReview questionReview = questionReviewRepository.findByIdAndUserId(reviewId, userId);
        questionReview.delete();
        questionReviewRepository.save(questionReview);
        applicationEventPublisher.publishEvent(DeletedReviewEvent.create(questionReview.getQuestionId(), questionReview.getRate()));
    }
}

File: qc-application/src/main/kotlin/com/eager/questioncloud/application/api/hub/review/service/HubReviewService.kt
package com.eager.questioncloud.application.api.hub.review.service

import com.eager.questioncloud.application.api.hub.review.dto.MyQuestionReview
import com.eager.questioncloud.application.api.hub.review.event.ReviewEvent
import com.eager.questioncloud.application.api.hub.review.event.ReviewEventType
import com.eager.questioncloud.application.api.hub.review.implement.*
import com.eager.questioncloud.core.common.PagingInformation
import com.eager.questioncloud.core.domain.review.dto.QuestionReviewDetail
import com.eager.questioncloud.core.domain.review.model.QuestionReview
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class HubReviewService(
    private val hubReviewReader: HubReviewReader,
    private val hubReviewRegister: HubReviewRegister,
    private val hubReviewUpdater: HubReviewUpdater,
    private val hubReviewRemover: HubReviewRemover,
    private val reviewEventProcessor: ReviewEventProcessor,
) {
    fun count(questionId: Long): Int {
        return hubReviewReader.count(questionId)
    }

    fun getQuestionReviewDetails(
        questionId: Long,
        userId: Long,
        pagingInformation: PagingInformation
    ): List<QuestionReviewDetail> {
        return hubReviewReader.getQuestionReviewDetails(questionId, userId, pagingInformation)
    }

    fun getMyQuestionReview(questionId: Long, userId: Long): MyQuestionReview {
        return hubReviewReader.getMyQuestionReview(questionId, userId)
    }

    @Transactional
    fun register(questionReview: QuestionReview) {
        hubReviewRegister.register(questionReview)
        reviewEventProcessor.saveEventLog(
            ReviewEvent.create(
                questionReview.questionId,
                questionReview.rate,
                ReviewEventType.REGISTER
            )
        )
    }

    @Transactional
    fun modify(reviewId: Long, userId: Long, comment: String, rate: Int) {
        val (questionId, varianceRate) = hubReviewUpdater.modify(reviewId, userId, comment, rate)
        reviewEventProcessor.saveEventLog(
            ReviewEvent.create(
                questionId,
                varianceRate,
                ReviewEventType.MODIFY
            )
        )
    }

    @Transactional
    fun delete(reviewId: Long, userId: Long) {
        val (questionId, varianceRate) = hubReviewRemover.delete(reviewId, userId)
        reviewEventProcessor.saveEventLog(
            ReviewEvent.create(
                questionId,
                varianceRate,
                ReviewEventType.DELETE
            )
        )
    }
}
